core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -489,8 +493,9 @@ class ReplicaManager(val config: KafkaConfig,
       val fetchPartitionStatus = logReadResults.map { case (topicAndPartition, result) =>
         (topicAndPartition, FetchPartitionStatus(result.info.fetchOffsetMetadata, fetchInfo.get(topicAndPartition).get))
       }
+
       val fetchMetadata = FetchMetadata(fetchMinBytes, fetchOnlyFromLeader, fetchOnlyCommitted, isFromFollower, fetchPartitionStatus)
-      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, responseCallback)
+      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota ,responseCallback)
```

Comment:
There could be a subtle issue with timeout. Say we check the quota and it's exceeded, and we will put the DelayedFetch in the purgatory. If no more bytes are produced, the DelayedFetch has to wait for maxWait. However, quota could become available before maxWait (and we won't get a chance to check). 

One potential way to address this is that if quota is exceeded, we calculate the amount of time that needs to pass before quota is available. We set the timeout in DelayedFetch to the be smaller of that time and maxWait. Then, if DelayedFetch expires, in DelayedFetch.onComplete(), if minBytes is not satisfied and maxWait hasn't been exceeded. We put DelayedFetch to purgatory again.